### PR TITLE
[orchestra] extract graph logic and enhance with cycle detection + coloring

### PR DESCRIPTION
**Pull request by drahnr**
* [x] extract graph logic
* [x] cycle detection (feature gated, `NetworkBridge` will always cause a cycle now)
* [x] highlight cycles (does not care about bounded or unbounded, that would require additional annotations)
* [x] add noone-sends and noone-consumes dummy nodes

### CODE REVIEW

**Comment by drahnr** [comment]
image

adds two extra nodes, unconsumed (skull) and never sent (fountain)

**Comment by drahnr** [comment]
Coloring cycles properly is pending, the logic is already there:

```
   Compiling orchestra v0.0.1 (/media/supersonic1t/projects/parity/polkadot/node/orchestra)
   Compiling polkadot-node-subsystem-types v0.9.23 (/media/supersonic1t/projects/parity/polkadot/node/subsystem-types)
   Compiling polkadot-overseer v0.9.23 (/media/supersonic1t/projects/parity/polkadot/node/overseer)
>>> Found 8 cycles <<< 
Wrote dot graph to /media/supersonic1t/projects/parity/polkadot/target/debug/build/orchestra-proc-macro-d0a92225d8323089/out/overseer-subsystem-messaging.dot
    Finished dev [unoptimized + debuginfo] target(s) in 4.66s
```

**Comment by drahnr** [comment]
image

with the output of:

```
⚡ Found 1 strongly connected cluster which includes at least one cycle
cycle(001) ∈ α: NetworkBridge ~~{"GossipSupportMessage"}~~> GossipSupport ~~{"NetworkBridgeMessage"}~~>  *
Wrote dot graph to /media/supersonic1t/projects/parity/polkadot/target/debug/build/orchestra-proc-macro-33f2b73e449a7593/out/overseer-subsystem-messaging.dot
```

(pre splitting the network bridge #5616 )

File: node/orchestra/proc-macro/src/graph.rs [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+						.edges_directed(node_idx, petgraph::Direction::Outgoing)
+						.find(|edge| edge.target() == node_idx)
+						.is_some()
+				},
+				0 => false,
+				_n => true,
+			}
+		}));
+		match sccs.len() {
+			0 => println!("✅ Found no strongly connected clusters, hence no cycles exist"),
```
Comment (ordian) [inline node/orchestra/proc-macro/src/graph.rs:124]: nit: a mixture in terminology "strongly connected component" vs "strongly connected cluster".
I haven't heard the latter, so the former is preferred: https://en.wikipedia.org/wiki/Strongly_connected_component.

File: node/orchestra/proc-macro/Cargo.toml [hunk 2]
```diff
@@ -28,10 +29,10 @@ thiserror = "1"
 tracing = "0.1"
 
 [features]
-default = []
+default = ["graph"]
```
Comment (ordian) [inline node/orchestra/proc-macro/Cargo.toml:32]: what was the reason to disable it by default? I'm not sure why need it by default

**Review by ordian — APPROVED** [review]

File: node/orchestra/proc-macro/Cargo.toml (hunk 2 above)
Comment (drahnr) [reply node/orchestra/proc-macro/Cargo.toml:32]: The reason was some issue with the CI system on `substrate` https://github.com/paritytech/ci_cd/issues/433 - I did not dig into this just yet.

File: node/orchestra/proc-macro/src/graph.rs (hunk 1 above)
Comment (drahnr) [reply node/orchestra/proc-macro/src/graph.rs:124]: Good catch :+1:

File: node/orchestra/proc-macro/Cargo.toml (hunk 2 above)
Comment (drahnr) [inline node/orchestra/proc-macro/Cargo.toml:32]:
```suggestion
default = [] # enable "graph" by default, blocked by <https://github.com/paritytech/ci_cd/issues/433>
```

**Review by sandreim — APPROVED** [review]
LGTM. It would also be really helpful to have the option to create a graph only with only the cycle.

**Comment by drahnr** [comment]
I'll create a separate PR that will add these graphs of cycles only

**Pull request merged**